src/main/scala/is/hail/methods/LinearMixedRegression.scala

```diff
@@ -161,150 +180,291 @@ object LinearMixedRegression {
     else
       vds2
   }
+
+  def computeNEigsDVF(S: DenseVector[Double], droppedVarianceFraction: Double): Int = {
+    require(0 <= droppedVarianceFraction && droppedVarianceFraction < 1)
+
+    val trace = sum(S)
+    var i = -1
+    var runningSum = 0.0
+    val target = droppedVarianceFraction * trace
+    while (runningSum <= target && i < S.length - 1) {
+      i += 1
+      //Note that S is increasing
+      runningSum += S(i)
+    }
+    S.length - i
+  }
+}
+
+trait ScalerLMM {
+  def likelihoodRatioTest(v: Vector[Double]): Annotation
 }
 
+// Handles full-rank case
+class FullRankScalerLMM(
+  y: DenseVector[Double],
+  yy: Double,
+  Qt: DenseMatrix[Double],
+  Qty: DenseVector[Double],
+  yQty: Double,
+  T: DenseMatrix[Double],
+  logNullS2: Double,
+  useML: Boolean) extends ScalerLMM {
+
+  val n = y.length
+  val invDf = 1.0 / (if (useML) n else n - Qt.rows)
+
+  def likelihoodRatioTest(x0: Vector[Double]): Annotation = {
+    val x = T * x0
+    val n = y.length
+    val Qtx = Qt * x
+    val xQtx: Double = (x dot x) - (Qtx dot Qtx)
+    val xQty: Double = (x dot y) - (Qtx dot Qty)
+
+    val b: Double = xQty / xQtx
+    val s2 = invDf * (yQty - xQty * b)
+    val chi2 = n * (logNullS2 - math.log(s2))
+    val p = chiSquaredTail(1, chi2)
+
+    Annotation(b, s2, chi2, p)
+  }
+}
+
+// Handles low-rank case, but is slower than ScalerLMM on full-rank case
+class LowRankScalerLMM(lmmConstants: LMMConstants, delta: Double, logNullS2: Double, useML: Boolean) extends ScalerLMM {
+  val n = lmmConstants.n
+  val d = lmmConstants.d
+  val k = lmmConstants.S.length
+  val y = lmmConstants.y
+  val covs = lmmConstants.C
+  val D = lmmConstants.S + delta
+  val Ut = lmmConstants.Ut
+  val Uty = lmmConstants.Uty
+  val UtCov = lmmConstants.UtC
+  val yty = lmmConstants.yty
+  val dy = Uty :/ D
+  val covtcov = lmmConstants.CtC
+  val covty = lmmConstants.Cty
+  val invDelta = 1 / delta
+  val Z = D.map(1 / _ - invDelta)
+  val ydy = yty / delta + (Uty dot (Uty :* Z))
+
+  val ZUtCov = UtCov(::, *) :* Z
+  val CzCLowerRight = UtCov.t * ZUtCov
+
+  def likelihoodRatioTest(x: Vector[Double]): Annotation = {
+    val xty = x dot y
+    val xtx = x dot x
+    val covtx = covs.t * x
+    val Utx = Ut * x
+    val UtC = DenseMatrix.horzcat(Utx.toDenseVector.toDenseMatrix.t, UtCov)
+
+    val CtC = DenseMatrix.zeros[Double](d + 1, d + 1)
```

Comment:
I think it's better to use `new DenseMatrix[Double](d + 1, d + 1)` here since the zeros are irrelevant